### PR TITLE
add narrowExtractedVectorUnaryOp to simplify cast nodes

### PR DESCRIPTION
**Pull request by vedantparanjape-amd**

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -24083,6 +24083,55 @@ static SDValue narrowInsertExtractVectorBinOp(SDNode *Extract,
...
+  unsigned NarrowingRatio = WideWidth / NarrowWidth;
+  unsigned WideNumElts = WideUVT.getVectorNumElements();
+
+  // Bail out if the target does not support a narrower version of the unaryop.
+  EVT NarrowUVT = EVT::getVectorVT(*DAG.getContext(), WideUVT.getScalarType(),
+                                   WideNumElts / NarrowingRatio);
+  if (!TLI.isOperationLegalOrCustomOrPromote(UnaryOpcode, NarrowUVT,
+                                             LegalOperations))
+    return SDValue();
+
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24126]: You need to add similar code to extract_subvector stage here: https://github.com/llvm/llvm-project/blob/170c525d79a4ab3659041b0655ac9697768fc915/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24157-L24169

Comment (vedantparanjape-amd) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24126]: Ah, I see. so there should a extract_subvec before fp_to_sint as well. I thought such a pattern would be already handled.

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24126]: What you're doing is fold extract_subvector(fp_to_sint(x),c) -> fp_to_sint(extract_subvector(x,c))

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -24123,7 +24123,30 @@ static SDValue narrowExtractedVectorUnaryOp(SDNode *Extract, SelectionDAG &DAG,
...
+    NarrowUEltVT.dump();
+    SDValue X = DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, NarrowUEltVT,
+                            UnaryOp.getOperand(0), Extract->getOperand(1));
+    SDValue NarrowUnaryOp = DAG.getNode(UnaryOpcode, DL, NarrowUVT, X);
+    X.dump();
+    NarrowUnaryOp.dump();
+    dbgs() << "reduced node found (smol)\n";
+    return DAG.getBitcast(VT, NarrowUnaryOp);
+  }
+
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24149]: Add `return SDValue()` here - if the above if() failed

Comment (vedantparanjape-amd) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24149]: Yes, did this. But it doesn't help with the crash.

**Comment by RKSimon** [comment]
Yes, you're right - we're getting stopped by legality tests as we have poor src/dst legality tests for cast ops.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 1 above)
Comment (vedantparanjape-amd) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24126]: ping! Thanks!

**Comment by vedantparanjape-amd** [comment]
> Yes, you're right - we're getting stopped by legality tests as we have poor src/dst legality tests for cast ops.

I see, is it something that can be worked on ?

**Comment by RKSimon** [comment]
> > Yes, you're right - we're getting stopped by legality tests as we have poor src/dst legality tests for cast ops.
> 
> I see, is it something that can be worked on ?

It isn't straightforward - cast ops require legality checks on the src/dst type pairs which isn't stored.

**Pull request closed without merging**